### PR TITLE
Refactor UPP data-fetching

### LINKED ISSUE #2501: Every 5 classifications, the classifier unmounts and remounts
## Package
app-project

## Describe the bug
https://frontend.preview.zooniverse.org/projects/eatyourgreens/-project-testing-ground/classify/workflow/3223?env=staging

Try submitting some classifications on that test Yes/No workflow. Watch today's session count in the 'Your Stats' box on the Classify page. At 5, 10, 15 classifications etc. the classifier is replaced by a loading icon and completely reloads.

## Expected behavior
`UserProjectReferences` should refresh in the background, every 5 classifications, but the classifier shouldn't completely reload, which forces a reload of the workflow and other resources from Panoptes.

### PR DESCRIPTION
**Pull request by eatyourgreens**
Fetch the full UPP resource when the stored user changes. When refreshing the UPP resource, fetch data in the background and only refresh stored settings, so as to avoid overwriting the stored classification count.

#2349 introduced a couple of undesired side effects for all projects:
- every 5 classifications, the user's `activity_count` is refreshed from Panoptes. `activity_count` is updated every 5 minutes, so this can lead to weird behaviour where your classification count, on the Classify page, can go up, then down, then up again.
- every 5 classifications, the classifier is replaced by a loading icon and reloads, including clearing the stored subject queue and refreshing it from Panoptes.

This PR should fix both those effects, by silently requesting UPP (every 5 classifications) and only refreshing `UPP.settings`.

Package:
app-project

Closes #2501.

# Review Checklist

## General

- [x] Are the tests passing locally and on Travis?
- [ ] Is the documentation up to date?

## Components
- [ ] Has a storybook story been created or updated?
- [ ] Is the component accessible? 
  - [ ] Can it be used with a screen reader? [BBC guide to testing with VoiceOver](https://bbc.github.io/accessibility-news-and-you/accessibility-and-testing-with-voiceover-os.html)
  - [ ] Can it be used from the keyboard? [WebAIM guide to keyboard testing](https://webaim.org/techniques/keyboard/#testing)
  - [ ] Is it passing accessibility checks in the storybook?

## Apps

- [x] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [x] Can you `yarn panic && yarn bootstrap` or `docker-compose up --build` and app works as expected?

## Publishing

- [ ] Is the changelog updated?
- [ ] Are the dependencies updated for apps and libraries that are using the newly published library?

## Post-merging

- [ ] Did the app deploy to https://frontend.preview.zooniverse.org/projects/:project-name/:owner or https://frontend.preview.zooniverse.org/about?
- [ ] Is the new feature working or bug now fixed?
  - [ ] Is there a Talk or blog post written to announce the new feature(s)?
- [ ] Is the design working across browsers (Firefox, Chrome, Edge, Safari) and mobile?
  - [ ] Is this approved by our designer?
- [ ] Is this ready for production deployment?

### CODE REVIEW

File: packages/app-project/stores/User/UserPersonalization/UserProjectPreferences/UserProjectPreferences.js [hunk 1]
```diff
@@ -97,13 +89,29 @@ const UserProjectPreferences = types
...
       refreshResource: flow(function * refreshResource() {
-        yield _fetch()
+        try {
+          const preferences = yield _fetch()
+          if (preferences) {
+            self.preferences = preferences.preferences
+          }
+        } catch (error) {
+          self.handleError(error)
+        }
```
Comment (eatyourgreens) [inline packages/app-project/stores/User/UserPersonalization/UserProjectPreferences/UserProjectPreferences.js:114]: This will crash the classifier if the refresh request fails (eg. I'm on mobile and temporarily lose network.) It might be better to ignore the error and let the volunteer continue to classify with their old UPP.

Comment (mcbouslog) [reply packages/app-project/stores/User/UserPersonalization/UserProjectPreferences/UserProjectPreferences.js:114]: I think it might be better to ignore the error and let the volunteer continue to classify.

**Review by mcbouslog — COMMENTED** [review]
Apologies, this took me some time to familiarize myself with. I confirmed the issue on staging project Planet Finders Beta, confirmed this branch fixes the issue.

A few questions for my understanding:
- why was `_fetch` refactored to remove `flow/yield` to `async/await`? was it because it's just requesting the project preference resource now, and not updating loading states at all (loading state updates moved to `fetchResource`, omitted from `refreshResource` to fix bug)
- this is outside of this PR, but I think important to confirm promotion still working as expected - I don't understand [the last test in UserProjectPreferences.spec.js](https://github.com/zooniverse/front-end-monorepo/blob/master/packages/app-project/stores/User/UserPersonalization/UserProjectPreferences/UserProjectPreferences.spec.js#L346-L368). 
    - should there be an assertion that equals `true`, like `expect(rootStore.user.personalization.projectPreferences.promptAssignment('123')).to.be.true()`, to test that if a user is on a different workflow than assigned they should be prompted to change to workflow assigned?
    - if the workflow the user's assigned(?) per `upp.settings.workflow_id` is '999' [code link](https://github.com/zooniverse/front-end-monorepo/blob/master/packages/app-project/stores/User/UserPersonalization/UserProjectPreferences/UserProjectPreferences.spec.js#L49), should that workflow be included in the test's project active workflows [code link](https://github.com/zooniverse/front-end-monorepo/blob/master/packages/app-project/stores/User/UserPersonalization/UserProjectPreferences/UserProjectPreferences.spec.js#L351), so an assertion could be made to confirm `promptAssignment` returns `true`?
    
I'll dig into the following a little more asap, but the reason not approving is that I just want to double-check if I'm seeing things correctly that on master I'm prompted again after 5 to try assigned workflow, but on this branch I don't think I am ever prompted again, on this testing project - https://local.zooniverse.org:3000/projects/srallen086/workflow-assignment-testing/classify/workflow/2333. Is that expected, and if so, should we create an issue to track how this breaks subsequent promotion prompting (but still merge this to fix the bug on all active projects, no active projects are using promotion (I think), so the no subsequent prompt issue can wait)?

**Comment by eatyourgreens** [comment]
That’s a good question about the tests. I’m not sure what they’re testing either. 

This was a refactor, so the behaviour of the store shouldn’t change. I’ll have another look and see if I can see what’s happening with workflow assignment. I’d expect that the new workflow is set when the preferences change, and then the front-end reacts to that change.

**Comment by eatyourgreens** [comment]
Reading back through the code, workflow assignment uses `settings`, not `preferences`. This has broken workflow assignment and the tests should be failing.

File: packages/app-project/stores/User/UserPersonalization/UserProjectPreferences/UserProjectPreferences.js [hunk 2]
```diff
@@ -97,13 +89,29 @@ const UserProjectPreferences = types
...
+          self.handleError(error)
         }
       }),
 
       refreshResource: flow(function * refreshResource() {
-        yield _fetch()
+        try {
+          const preferences = yield _fetch()
+          if (preferences) {
+            self.preferences = preferences.preferences
```
Comment (eatyourgreens) [inline packages/app-project/stores/User/UserPersonalization/UserProjectPreferences/UserProjectPreferences.js:110]: This should update `self.settings`.

Comment (eatyourgreens) [reply packages/app-project/stores/User/UserPersonalization/UserProjectPreferences/UserProjectPreferences.js:110]:
```suggestion
            self.settings = preferences.settings
```

**Comment by eatyourgreens** [comment]
@mcbouslog I’ve now vanished down a rabbit hole of trying to figure out how the tests work. I think you’re right, that they pass when they should fail, so there’s an evergreen test in there somewhere. 

I think there’s also a bunch of old code in the setup that isn’t needed, now that the `User` model has been refactored to remove some of the `autorun` observers. I can clean up old code here, but the evergreen tests might be worth their own issue.

**Comment by eatyourgreens** [comment]
Sorry, forgot to say that I would prioritise fixing the classifier for live projects over breaking workflow assignment, if workflow assignment isn’t live. You’re right that we can circle back to fix that when it’s needed.

**Comment by eatyourgreens** [comment]
There were a couple of problems with the tests that I've started to fix here:
- `rootStore` was defined in the top level scope, so that store state leaked across tests. I've moved it down to the scope of each individual test block. That means more setup code but loading state etc. is properly reset for each set of tests.
- tests weren't waiting for the mock UPP data fetching to complete. I've added MobX's [`when`](https://www.mobxjs.com/refguide/when.html) to run the tests when loading is complete. That technique is more-or-less copied straight from [the classifier's `UPPStore` tests](https://github.com/zooniverse/front-end-monorepo/blob/master/packages/lib-classifier/src/store/UserProjectPreferencesStore.spec.js).

I'm still not sure what the workflow assignment tests should be testing, so I've left those out of this PR.

**Comment by eatyourgreens** [comment]
@mcbouslog I added logging, and here's why those tests are broken:
```
Stores > UserProjectPreferences
    Views > promptAssignment
      when a workflow is not currently selected
{ assignedWorkflowID: '555', currentWorkflowID: undefined }
        ✓ should not prompt the user
      when the assigned workflow is not active
{ assignedWorkflowID: undefined, currentWorkflowID: '123' }
        ✓ should not prompt the user
      when the assigned workflow is the same as the current workflow
{ assignedWorkflowID: undefined, currentWorkflowID: '555' }
        ✓ should not prompt the user
      when the assigned workflow is not the same as the current workflow
{ assignedWorkflowID: undefined, currentWorkflowID: '123' }
        ✓ should prompt the user
```

**Comment by eatyourgreens** [comment]
I think all that's left to do here is to write some tests that refresh the assigned workflow without unexpected changes to the user classification count or classifier loading state.

**Comment by eatyourgreens** [comment]
I'm not sure how to test workflow assignment on the test project, but I've added some tests that refresh your settings and they update the assigned workflow ID without changing your classification count or the classifier loading state.

File: packages/app-project/stores/User/UserPersonalization/UserProjectPreferences/UserProjectPreferences.js (hunk 1 above)
Comment (eatyourgreens) [reply packages/app-project/stores/User/UserPersonalization/UserProjectPreferences/UserProjectPreferences.js]: I've changed this to log the error without changing the loading state.

**Review by mcbouslog — APPROVED** [review]
This looks great! Confirmed this fixes the classifier reloading issue and stats resetting issue. 

tl;dr - this breaks prompting every 5 classifications if promoted to a different workflow than the workflow classifying on, but that's ok - that needs further refactoring and still prompts initially. 

longer story - This still requests and updates user project preferences (`settings.workflow_id`) every five classifications to see if a user is promoted for the workflow promotion feature, however, it breaks the prompting every 5 classifications because the WorkflowAssignmentModalContainer is using `useEffect` based on `loadingState` which is no longer updated on settings updates. I think fixing the subsequent prompt should be revisited in subsequent work. We could change the WorkflowAssignmentModalContainer `useEffect` to listen to `assignedWorkflowID` and then I think it would prompt only if promoted. That might be a nice quick fix for a subsequent PR. Or as discussed in standup dig into using Sugar, which would be a better longterm solution (then can remove requests every 5 classifications).
